[CI] IndexRecoveryIT.testRerouteRecovery : Paths exist that should have been deleted

Build failures: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.4+matrix-java-periodic/ES_BUILD_JAVA=java10,ES_RUNTIME_JAVA=java10,nodes=virtual&&linux/21/console

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+matrix-java-periodic/ES_BUILD_JAVA=java10,ES_RUNTIME_JAVA=java11,nodes=virtual&&linux/224/console

Root cause:
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=5109, name=elasticsearch[node_t0][clusterApplierService#updateTask][T#1], state=RUNNABLE, group=TGRP-IndexRecoveryIT]
	at __randomizedtesting.SeedInfo.seed([E8CEC7E4E4750104:456574B0D12D6346]:0)
Caused by: java.lang.AssertionError: Paths exist that should have been deleted: [/var/lib/jenkins/workspace/elastic+elasticsearch+6.4+matrix-java-periodic/ES_BUILD_JAVA/java10/ES_RUNTIME_JAVA/java10/nodes/virtual&&linux/server/build/testrun/integTest/J1/temp/org.elasticsearch.indices.recovery.IndexRecoveryIT_E8CEC7E4E4750104-001/tempDir-004/data/nodes/0/indices/RJoRvWIVRDyM1toTfMV2tw/0]
	at __randomizedtesting.SeedInfo.seed([E8CEC7E4E4750104]:0)
	at org.elasticsearch.env.NodeEnvironment.assertPathsDoNotExist(NodeEnvironment.java:469)
	at org.elasticsearch.env.NodeEnvironment.deleteShardDirectoryUnderLock(NodeEnvironment.java:459)
	at org.elasticsearch.env.NodeEnvironment.deleteShardDirectorySafe(NodeEnvironment.java:399)
	at org.elasticsearch.indices.IndicesService.deleteShardStore(IndicesService.java:770)
	at org.elasticsearch.indices.store.IndicesStore$ShardActiveResponseHandler.lambda$allNodesResponded$2(IndicesStore.java:288)
	at org.elasticsearch.cluster.service.ClusterApplierService.lambda$runOnApplierThread$0(ClusterApplierService.java:306)
	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.apply(ClusterApplierService.java:155)
	at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:399)
	at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:160)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:624)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:244)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:207)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
```

It does not reproduce:
```
REPRODUCE WITH: ./gradlew :server:integTest \
  -Dtests.seed=B4F936D0EF2B4D07 \
  -Dtests.class=org.elasticsearch.indices.recovery.IndexRecoveryIT \
  -Dtests.method="testRerouteRecovery" \
  -Dtests.security.manager=true \
  -Dtests.locale=fr-TN \
  -Dtests.timezone=Eire
```

A failure occurred in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+multijob-unix-compatibility/os=oraclelinux/38/console that suggests #34120 doesn't fix this in every case.

The error is this:

```
06:45:17 ERROR   0.76s J1 | IndicesLifecycleListenerIT.testIndexStateShardChanged <<< FAILURES!
06:45:17    > Throwable #1: com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=10795, name=elasticsearch[node_t0][clusterApplierService#updateTask][T#1], state=RUNNABLE, group=TGRP-IndicesLifecycleListenerIT]
06:45:17    > 	at __randomizedtesting.SeedInfo.seed([A8D19CF9DD9EFD43:7E04533A46A0D37A]:0)
06:45:17    > Caused by: java.lang.AssertionError: Paths exist that should have been deleted: [/var/lib/jenkins/workspace/elastic+elasticsearch+6.x+multijob-unix-compatibility/os/oraclelinux/server/build/testrun/integTest/J1/temp/org.elasticsearch.indices.IndicesLifecycleListenerIT_A8D19CF9DD9EFD43-001/tempDir-002/d1/nodes/0/indices/2HM7dFt2TTqo3nNIgS8seA/2, /var/lib/jenkins/workspace/elastic+elasticsearch+6.x+multijob-unix-compatibility/os/oraclelinux/server/build/testrun/integTest/J1/temp/org.elasticsearch.indices.IndicesLifecycleListenerIT_A8D19CF9DD9EFD43-001/tempDir-002/d0/nodes/0/indices/2HM7dFt2TTqo3nNIgS8seA/2]
```

Looking elsewhere in the log file it seems that `2HM7dFt2TTqo3nNIgS8seA/2` _is_ the state directory, as there are messages about `.../2HM7dFt2TTqo3nNIgS8seA/2/_state/state-1.st`.

The repro command for this failure was:

```
./gradlew :server:integTest \
  -Dtests.seed=A8D19CF9DD9EFD43 \
  -Dtests.class=org.elasticsearch.indices.IndicesLifecycleListenerIT \
  -Dtests.method="testIndexStateShardChanged" \
  -Dtests.security.manager=true \
  -Dtests.locale=es-AR \
  -Dtests.timezone=America/Toronto \
  -Dcompiler.java=10 \
  -Druntime.java=8
```

This didn't reproduce locally.

--- a/core/src/main/java/org/elasticsearch/env/NodeEnvironment.java
+++ b/core/src/main/java/org/elasticsearch/env/NodeEnvironment.java
@@ -59,6 +59,7 @@ import org.elasticsearch.node.Node;
 
 import java.io.Closeable;
 import java.io.IOException;
+import java.io.UncheckedIOException;
 import java.nio.file.AtomicMoveNotSupportedException;
 import java.nio.file.DirectoryStream;
 import java.nio.file.FileStore;
@@ -70,6 +71,7 @@ import java.util.Arrays;
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
@@ -78,6 +80,8 @@ import java.util.Set;
 import java.util.concurrent.Semaphore;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import static java.util.Collections.unmodifiableSet;
 
@@ -513,7 +517,33 @@ public final class NodeEnvironment  implements Closeable {
             IOUtils.rm(customLocation);
         }
         logger.trace("deleted shard {} directory, paths: [{}]", shardId, paths);
-        assert FileSystemUtils.exists(paths) == false;
+        assert assertPathsDoNotExist(paths);
+    }
+
+    private static boolean assertPathsDoNotExist(final Path[] paths) {
+        Set<Path> existingPaths = Stream.of(paths)
+            .filter(FileSystemUtils::exists)
+            .filter(leftOver -> {
+                // Relaxed assertion for the special case where only the empty state directory exists after deleting
+                // the shard directory because it was created again as a result of a metadata read action concurrently.
+                try (DirectoryStream<Path> children = Files.newDirectoryStream(leftOver)) {
+                    Iterator<Path> iter = children.iterator();
+                    if (iter.hasNext() == false) {
+                        return true;
+                    }
+                    Path maybeState = iter.next();
+                    if (iter.hasNext() || maybeState.equals(leftOver.resolve(MetaDataStateFormat.STATE_DIR_NAME)) == false) {
+                        return true;
+                    }
+                    try (DirectoryStream<Path> stateChildren = Files.newDirectoryStream(maybeState)) {
+                        return stateChildren.iterator().hasNext();
+                    }
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }).collect(Collectors.toSet());
+        assert existingPaths.size() == 0 : "Paths exist that should have been deleted: " + existingPaths;
+        return existingPaths.size() == 0;
     }
 
     private boolean isShardLocked(ShardId id) {

